### PR TITLE
task 4 complete and ready for review

### PR DESCRIPTION
**Pull request by jxnbox**

### CODE REVIEW

File: controllers.js [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
     })
-    .catch(next)
+    .catch(next);
+}
 
+exports.getArticle = (req, res, next) => {
+    getArticleModel()
+    .then((articles) => {
+        console.log(articles)
+        res.status(200).send(articles);
```
Comment (popatre) [inline controllers.js]: You need to return an object, with a key of articles. 
This is the same for your controller above, but `topics`

File: controllers.js [hunk 2]
```diff
@@ -1,10 +1,21 @@
...
         res.status(200).send(topics);
     })
-    .catch(next)
+    .catch(next);
+}
 
+exports.getArticle = (req, res, next) => {
+    getArticleModel()
+    .then((articles) => {
+        console.log(articles)
```
Comment (popatre) [inline controllers.js]: remove console log

File: __tests__/app.test.js [hunk 3]
```diff
@@ -41,4 +41,44 @@ describe('1. GET api topics', () => {
...
+})
+
+describe('2. GET api articles', () => {
+    describe('a. status 200 & Data', () => {
+        it('returns 200 status code & an array back to the user', () => {
+            return request(app)
+                .get('/api/articles')
+                .expect(200)
+                .then( (res) => {
+                    const articles = res.body;
```
Comment (popatre) [inline __tests__/app.test.js]: Will need to be returned on a key of articles

File: __tests__/app.test.js [hunk 4]
```diff
@@ -41,4 +41,44 @@ describe('1. GET api topics', () => {
...
+                        expect(article).toEqual(
+                            expect.objectContaining({
+                                article_id : expect.any(Number),
+                                title : expect.any(String),
+                                topic : expect.any(String),
+                                author : expect.any(String),
+                                body : expect.any(String),
+                                created_at : expect.any(String),
+                                votes : expect.any(Number)
+                            })
```
Comment (popatre) [inline __tests__/app.test.js]: The ticket asks for:

comment_count which is the total count of all the comments with this article_id - you should make use of queries to the database in order to achieve this.

It also doesn't want the body

File: __tests__/app.test.js [hunk 5]
```diff
@@ -41,4 +41,44 @@ describe('1. GET api topics', () => {
...
+                        )
+                    })
+                })
+        })
+    })
```
Comment (popatre) [inline __tests__/app.test.js:73]: Ticket also asks for:

- [ ]  the articles should be sorted by date in descending order.

Could you add a test and the code for this?

File: __tests__/app.test.js [hunk 6]
```diff
@@ -41,4 +41,44 @@ describe('1. GET api topics', () => {
...
+                                votes : expect.any(Number)
+                            })
+                        )
+                    })
+                })
+        })
+    })   
+    
+    describe('b. error handling', () => {
+        it('send a 404 status code when user inputs a bad path', () => {
```
Comment (popatre) [inline __tests__/app.test.js:76]: nice test here

File: controllers.js [hunk 7]
```diff
@@ -1,10 +1,21 @@
-const {getTopicModel} = require('./models')
+const {
+    getTopicModel,
+    getArticleModel
+} = require('./models')
 
 exports.getTopic = (req, res, next) => {
     getTopicModel()
     .then((topics) => {
         res.status(200).send(topics);
```
Comment (popatre) [inline controllers.js:9]: send back an object with this key

File: models.js [hunk 8]
```diff
@@ -3,6 +3,13 @@ const db = require('./db/connection')
 exports.getTopicModel = () => {
     return db.query('SELECT * FROM topics;')
         .then( (topics) => {
-            return topics.rows
+            return topics.rows;
+        })
+}
+
+exports.getArticleModel = () => {
+    return db.query('SELECT * FROM articles;')
```
Comment (popatre) [inline models.js]: Check the ticket for what it wants to be sent back, it doesnt want all the keys, and wants a comment count

**Review by popatre — CHANGES_REQUESTED** [review]

File: models.js [hunk 9]
```diff
@@ -1,8 +1,34 @@
...
+        db.query('SELECT article_id FROM comments;')
+        .then( (articles) => {
+            return articles.rows;
+        })
+    ])
+    .then( (promiseArr) => {
+        promiseArr[0].forEach(article => {
+            article.comment_count = 0;
+            promiseArr[1].forEach(comment => {
+                if (article.article_id === comment.article_id) {
```
Comment (Pipe-Smoking-Rabbit) [inline models.js]: it's impressive that you've managed to achieve a comment count through javascript - but this is more easily achievable by just using SQL making use of `COUNT` and `JOIN` functionality 

I would say that you're missing an opportunity here to showcase SQL skills to employers, and so I would recommend refactoring this to just use one single SQL query that joins and creates a `comment_count` with the `COUNT` function

File: controllers.js [hunk 10]
```diff
@@ -1,10 +1,20 @@
...
     })
-    .catch(next)
+    .catch(next);
+}
 
+exports.getArticle = (req, res, next) => {
+    getArticleModel()
+    .then((articles) => {
+        res.status(200).send(articles);
```
Comment (Pipe-Smoking-Rabbit) [inline controllers.js]: unfortunately this does still need to be changed - the response should be an object with a key containing the data, i.e. `.send({ articles: articles });`  which can be shortened to  `.send({ articles });`

File: __tests__/app.test.js [hunk 11]
```diff
@@ -41,4 +41,47 @@ describe('1. GET api topics', () => {
...
+                                author : expect.any(String),
+                                body : expect.any(String),
+                                created_at : expect.any(String),
+                                votes : expect.any(Number),
+                                comment_count : expect.any(Number)
+                            })
+                        )
+                    })
+                    expect(articles[0].article_id).toBe(3);
+                    expect(articles[articles.length - 1].article_id).toBe(7);
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js]: this test is great! 

We're missing some more testing though - we need to check that the data is coming back sorted by `created_at` in `desc` order as per the ticket instructions 

sounds like a job for Jest-Sorted to me

**Review by Pipe-Smoking-Rabbit — CHANGES_REQUESTED** [review]
There's a bit more work to be done on this ticket, it's coming along really well but just have a check over the recent comments to see what else you need to include!

File: __tests__/app.test.js [hunk 12]
```diff
@@ -63,12 +63,11 @@ describe('2. GET api articles', () => {
...
                                 created_at : expect.any(String),
                                 votes : expect.any(Number),
-                                comment_count : expect.any(Number)
+                                comment_count : expect.any(String)
                             })
                         )
                     })
-                    expect(articles[0].article_id).toBe(3);
-                    expect(articles[articles.length - 1].article_id).toBe(7);
+                    expect(articles).toBeSortedBy('created_at', {descending : true});
```
Comment (robert112233) [inline __tests__/app.test.js:70]: toBeSortedBy is great to check sort order, in future you could extract this out to another test. It is good practice when using TDD to have separate tests to test individual behaviours of the function. Not a huge issue though you can still merge 👍

**Review by robert112233 — APPROVED** [review]
Merge away 👍

**Pull request merged**